akka-actor-typed/src/main/scala/akka/actor/typed/eventstream/EventStream.scala

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (C) 2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.actor.typed.eventstream
+
+import akka.actor.typed._
+import akka.actor.typed.internal.adapter.EventStreamAdapter
+import akka.actor.typed.scaladsl.adapter._
+import akka.annotation.{ DoNotInherit, InternalApi }
+
+import scala.reflect.ClassTag
+
+/**
+ * INTERNAL API
+ *
+ * Exposes a typed actor that interacts with the [[akka.actor.ActorSystem.eventStream]].
+ *
+ * It is used as an extension to ensure a single instance per actor system.
+ */
+@InternalApi private[akka] final class EventStream(actorSystem: ActorSystem[_]) extends Extension {
+  val ref: ActorRef[EventStream.Command] =
+    actorSystem.internalSystemActorOf(EventStreamAdapter.behavior, "eventstream", Props.empty)
+}
+
+object EventStream extends ExtensionId[EventStream] {
+
+  /**
+   * Not for user Extension
+   */
+  @DoNotInherit sealed trait Command
+
+  /**
+   * Publish an event of type E
+   * @param event
+   * @tparam E
+   */
+  final case class Publish[E](event: E) extends Command
+
+  /**
+   * Subscribe a typed actor to listen for types or subtypes of E.
+   * ==Simple example==
+   * {{{
+   *   sealed trait A
+   *   case object A1 extends A
+   *   //listen for all As
+   *   def subscribe(actorSystem: ActorSystem[_], actorRef: ActorRef[A]) =
+   *     actorSystem.eventStream ! Subscribe(actorRef)
+   *   //listen for A1s only
+   *   def subscribe(actorSystem: ActorSystem[_], actorRef: ActorRef[A]) =
+   *     actorSystem.eventStream ! Subscribe[A1](actorRef)
+   * }}}
+   *
+   * @param subscriber
+   * @param classTag
+   * @tparam E
+   */
+  final case class Subscribe[E](subscriber: ActorRef[E])(implicit classTag: ClassTag[E]) extends Command {
+    private[akka] def topic: Class[_] = classTag.runtimeClass
+  }
+
+  /**
+   * Unsubscribe an actor ref from the event stream
+   * @param subscriber
+   * @tparam E
+   */
+  final case class Unsubscribe[E](subscriber: ActorRef[E]) extends Command
+
+  override def createExtension(system: ActorSystem[_]): EventStream = new EventStream(system)
+
+  /**
+   * Java API.
+   */
+  def subscribe[E](subscriber: ActorRef[E], clazz: java.lang.Class[E]): Subscribe[E] =
```

Comment:
this can be confusing, not very clear that this is just a factory method for creating the `Subscribe` message. Maybe we can just add another constructor to `Subscribe` and the Java API can be `new Subscribe(subscriber, clazz)`?

Reply:
Ok I can do this constructor. Would it be also a good idea to have a smart constructor or do you think it's redundant ? Most Java APIs now have smart constructors (factory methods) in search for better semantics. I find it way friendlier to do `EventStream.subscribe()` than use new. Maybe we need to find a better name than `subscribe`, e.g. subscribeToType ?

Reply:
I added both, a factory method and the constructor, if you prefer I can remove the factory method